### PR TITLE
:book: Updating doc with Installation steps of cert-manager and prometheus

### PR DESCRIPTION
**Pull request by yashsingh74**
This PR aims to update the documentation with the Installation steps of cert-manager and prometheus. 
This will helps to achieve the running of the webhooks in the test server.

Fixes: #2777

### CODE REVIEW

File: docs/book/src/reference/envtest.md [hunk 1]
```diff
@@ -151,6 +151,16 @@ testEnv = &envtest.Environment{
 }
 ```
 
+## Installing the cert-manager
+
+To test the webhooks, `cert-manager` needs to be installed in the test server. This will help for provisioning the certificates for the webhook server.
+You can follow [the cert manager documentation](https://cert-manager.io/docs/installation/) to install it.
```
Comment (camilamacedo86) [inline docs/book/src/reference/envtest.md:157]: @yashsingh74, see that the envtest has options for you set up the configs: https://pkg.go.dev/sigs.k8s.io/controller-runtime/pkg/envtest#WebhookInstallOptions.

So, I believe that we need to use the options to set it up. Also, would be nice if we could add some examples here and to link the go dev docs for further info. WDYT?

Comment (yashsingh74) [reply docs/book/src/reference/envtest.md]: @camilamacedo86 Please check, I hope you are referring to this change ?

File: docs/book/src/reference/envtest.md [hunk 2]
```diff
@@ -96,13 +96,15 @@ You can use environment variables and/or flags to specify the `kubectl`,`api-ser
...
+| `KUBEBUILDER_CONTROLPLANE_START_TIMEOUT` and `KUBEBUILDER_CONTROLPLANE_STOP_TIMEOUT` | durations in format supported by [`time.ParseDuration`](https://golang.org/pkg/time/#ParseDuration) | Specify timeouts different from the default for the test control plane to (respectively) start and stop; any test run that exceeds them will fail.                                                                                                                   |
+| `KUBEBUILDER_ATTACH_CONTROL_PLANE_OUTPUT`                                 | boolean | Set to `true` to attach the control plane's stdout and stderr to os.Stdout and os.Stderr. This can be useful when debugging test failures, as output will include output from the control plane.                                                                     |
+ | `TEST_ASSET_CERT_MANAGER`                                                 | path to directory | Point integration tests to a directory containing cert-manager binaries. Follow [cert-manager documentation](https://cert-manager.io/docs/installation/) to install it.                                                                                              |
```
Comment (camilamacedo86) [inline docs/book/src/reference/envtest.md:106]: @yashsingh74, 

Did you test these options `TEST_ASSET_CERT_MANAGER` and `TEST_ASSET_PROMETHEUS`? I cannot find these ones in the controller-runtime implementation. Where do you check that env vars can be used to set it for the env test?

Comment (camilamacedo86) [reply docs/book/src/reference/envtest.md:106]: Also, did you check the envtest config options for webhooks https://pkg.go.dev/sigs.k8s.io/controller-runtime/pkg/envtest#WebhookInstallOptions? Did you try to use them to config the cert and etc?

Comment (camilamacedo86) [reply docs/book/src/reference/envtest.md:106]: Hi @yashsingh74,  

By looking at the issue https://github.com/kubernetes-sigs/kubebuilder/issues/2777 (it was opened too long ago and it is might out of context nowadays), I think that we can add here:

See that the envtest is commonly used to test the controllers.It is an integration test to "unit" test the controller/reconcile implementation. 

Therefore, we do not need use the metrics or cert-manager to run these tests usually. For example, look: https://github.com/kubernetes-sigs/kubebuilder/blob/master/testdata/project-v3-with-deploy-image/controllers/busybox_controller_test.go

However, if you want to run the tests against a cluster where you have some configurations in place they can use the option to test using an existing cluster:

```golang
                testEnv = &envtest.Environment{
			UseExistingCluster: true,
		}
```

Note that we can link/share the ENV config options: https://pkg.go.dev/sigs.k8s.io/controller-runtime/pkg/envtest#Environment

Also, we can let the users know that by using the plugin deploy-image/v1-alpha they can check an example of the code generated to use env tests to test the controller implementation via a note. See:
- https://book.kubebuilder.io/quick-start.html#using-deploy-image-plugin-to-generate-apis-and-controllers-code
- https://github.com/kubernetes-sigs/kubebuilder/blob/master/testdata/project-v3-with-deploy-image/controllers/busybox_controller_test.go

On top of that, share that if they are using webhooks they might want to check the optional config options https://pkg.go.dev/sigs.k8s.io/controller-runtime/pkg/envtest#WebhookInstallOptions

File: docs/book/src/reference/envtest.md [hunk 3]
```diff
@@ -97,12 +97,12 @@ You can use environment variables and/or flags to specify the `kubectl`,`api-ser
 ### Environment Variables
```
Comment (camilamacedo86) [inline docs/book/src/reference/envtest.md:97]: Under Writing test, at the bottom of the section see: https://github.com/kubernetes-sigs/kubebuilder/blob/7487261ff6bf915b936f47684b2ffe1b23e2d689/docs/book/src/reference/envtest.md#writing-tests

We can add the following note like:

```
<aside class="note">
<h1>Examples</h1>

You can use the plugin [DeployImage](https://book.kubebuilder.io/plugins/deploy-image-plugin-v1-alpha.html) to check examples. This plugin allows users to scaffold API/Controllers to deploy and manage an Operand (image) on the cluster following the guidelines and best practices. It abstracts the complexities of achieving this goal while allowing users to customize the generated code.

Therefore, you can check that a test using ENV TEST will be generated for the controller which has the purpose to ensure that the Deployment is created succeffully. You can see an example of its code implementation under the `testdata` directory with the [DeployImage](https://book.kubebuilder.io/plugins/deploy-image-plugin-v1-alpha.html) samples [here](https://github.com/kubernetes-sigs/kubebuilder/blob/v3.7.0/testdata/project-v3-with-deploy-image/controllers/busybox_controller_test.go).

</aside>
```

File: docs/book/src/reference/envtest.md [hunk 4]
```diff
@@ -97,12 +97,12 @@ You can use environment variables and/or flags to specify the `kubectl`,`api-ser
...
 
 | Variable name | Type | When to use |
```
Comment (camilamacedo86) [inline docs/book/src/reference/envtest.md:99]: Also, regards the links we need to use an alias and add them at the bottom of the page for better maintainability. Example:

[text][alias]

[alias]: link -> this one is at the bottom of the file.

File: docs/book/src/reference/envtest.md [hunk 5]
```diff
@@ -97,12 +97,12 @@ You can use environment variables and/or flags to specify the `kubectl`,`api-ser
...
-| `KUBEBUILDER_CONTROLPLANE_START_TIMEOUT` and `KUBEBUILDER_CONTROLPLANE_STOP_TIMEOUT` | durations in format supported by [`time.ParseDuration`](https://golang.org/pkg/time/#ParseDuration) | Specify timeouts different from the default for the test control plane to (respectively) start and stop; any test run that exceeds them will fail. |
-| `KUBEBUILDER_ATTACH_CONTROL_PLANE_OUTPUT` | boolean | Set to `true` to attach the control plane's stdout and stderr to os.Stdout and os.Stderr. This can be useful when debugging test failures, as output will include output from the control plane. |
+| --- | :--- | :---                                                                                                                                                                                                                                                    |
+| `USE_EXISTING_CLUSTER` | boolean | Instead of setting up a local control plane, point to the control plane of an existing cluster.                                                                                                                                                         |
```
Comment (camilamacedo86) [inline docs/book/src/reference/envtest.md:101]:
```suggestion                                                                                                                                       |
```
I did not check in the C+R code but I am not sure if we have this env var config implemented there. How to use this option is described in the new section. Anyway, I do not think that we should add this one here since it can brings confusion because the section has the purpose to explain its config setup.

File: docs/book/src/reference/envtest.md [hunk 6]
```diff
@@ -169,3 +169,17 @@ Expect(deployment.ObjectMeta.OwnerReferences).To(ContainElement(expectedOwnerRef
 
 [envtest]:https://pkg.go.dev/sigs.k8s.io/controller-runtime/pkg/envtest
 [setup-envtest]:https://pkg.go.dev/sigs.k8s.io/controller-runtime/tools/setup-envtest
```
Comment (camilamacedo86) [inline docs/book/src/reference/envtest.md:171]: These ones are alias and the links should be kept at the end of the page.

File: docs/book/src/reference/envtest.md [hunk 7]
```diff
@@ -116,7 +116,7 @@ var _ = BeforeSuite(func(done Done) {
 
 	logf.SetLogger(zap.New(zap.WriteTo(GinkgoWriter), zap.UseDevMode(true)))
 	testenv = &envtest.Environment{}
-
```
Comment (camilamacedo86) [inline docs/book/src/reference/envtest.md:119]: Before the flags section we can add a NOTE like, wdyt?

````
<aside class="note">
<h1>ENV TEST Config Options</h1>

You can look at the controller-runtime docs to know more about its configuration options, see [here](https://pkg.go.dev/sigs.k8s.io/controller-runtime/pkg/envtest#Environment)

</aside>
````
UseExistingCluster

File: docs/book/src/reference/envtest.md [hunk 8]
```diff
@@ -133,6 +141,13 @@ var _ = AfterSuite(func() {
 })
 ```
 
+<aside class="note">
+<h1>ENV TEST Config Options</h1>
+
+You can look at the controller-runtime docs to know more about its configuration options, see [here](https://pkg.go.dev/sigs.k8s.io/controller-runtime/pkg/envtest#Environment)
```
Comment (camilamacedo86) [inline docs/book/src/reference/envtest.md:147]:
```suggestion
You can look at the controller-runtime docs to know more about its configuration options, see [here](https://pkg.go.dev/sigs.k8s.io/controller-runtime/pkg/envtest#Environment). On top of that, if you are
looking to use ENV TEST to test your webhooks then you might want to give a look at its install [options](https://pkg.go.dev/sigs.k8s.io/controller-runtime/pkg/envtest#WebhookInstallOptions).  
```

File: docs/book/src/reference/envtest.md [hunk 9]
```diff
@@ -167,5 +183,32 @@ expectedOwnerReference := v1.OwnerReference{
...
+
+```go
+BeforeEach(func() {
+	By("installing the cert-manager bundle")
+	Expect(kbc.InstallCertManager(false)).To(Succeed())
+
+	By("installing the Prometheus operator")
+	Expect(kbc.InstallPrometheusOperManager()).To(Succeed())
+})
+```
```
[the remark this answers is not in the document]
Comment (camilamacedo86) [reply docs/book/src/reference/envtest.md:201]: BEcause the links use Kubebuilder LIB for e2e tests to help out to mock things on the e2e tests done in the repo which are not good for these ones.

File: docs/book/src/reference/envtest.md [hunk 10]
```diff
@@ -167,5 +183,43 @@ expectedOwnerReference := v1.OwnerReference{
...
+BeforeEach(func() {
+	By("installing prometheus operator")
+	Expect(utils.InstallPrometheusOperator()).To(Succeed())
+	
+	By("installing the cert-manager")
+	Expect(utils.InstallCertManager()).To(Succeed())
+}
+
+// You can also remove them after the tests::
+AfterEach(func() {
```
Comment (camilamacedo86) [inline docs/book/src/reference/envtest.md:205]:
```suggestion
 	// You can also remove them after the tests::
 	AfterEach(func() {
```

File: docs/book/src/reference/envtest.md [hunk 11]
```diff
@@ -167,5 +183,43 @@ expectedOwnerReference := v1.OwnerReference{
...
+	By("uninstalling the Prometheus manager bundle")
+	utils.UninstallPrometheusOperManager()
+	
+	By("uninstalling the cert-manager bundle")
+	utils.UninstallCertManager()
+})
+```
+
```
Comment (camilamacedo86) [inline docs/book/src/reference/envtest.md:213]:
```suggestion

Check the following example of how you can implement the above operations:

```go
 const (
	prometheusOperatorVersion = "0.51"
	prometheusOperatorURL     = "https://raw.githubusercontent.com/prometheus-operator/" +
		"prometheus-operator/release-%s/bundle.yaml"

	certmanagerVersion = "v1.5.3"
	certmanagerURLTmpl = "https://github.com/jetstack/cert-manager/releases/download/%s/cert-manager.yaml"
)

func warnError(err error) {
	fmt.Fprintf(GinkgoWriter, "warning: %v\n", err)
}

// InstallPrometheusOperator installs the prometheus Operator to be used to export the enabled metrics.
func InstallPrometheusOperator() error {
	url := fmt.Sprintf(prometheusOperatorURL, prometheusOperatorVersion)
	cmd := exec.Command("kubectl", "apply", "-f", url)
	_, err := Run(cmd)
	return err
}

// UninstallPrometheusOperator uninstalls the prometheus
func UninstallPrometheusOperator() {
	url := fmt.Sprintf(prometheusOperatorURL, prometheusOperatorVersion)
	cmd := exec.Command("kubectl", "delete", "-f", url)
	if _, err := Run(cmd); err != nil {
		warnError(err)
	}
}

// UninstallCertManager uninstalls the cert manager
func UninstallCertManager() {
	url := fmt.Sprintf(certmanagerURLTmpl, certmanagerVersion)
	cmd := exec.Command("kubectl", "delete", "-f", url)
	if _, err := Run(cmd); err != nil {
		warnError(err)
	}
}

// InstallCertManager installs the cert manager bundle.
func InstallCertManager() error {
	url := fmt.Sprintf(certmanagerURLTmpl, certmanagerVersion)
	cmd := exec.Command("kubectl", "apply", "-f", url)
	if _, err := Run(cmd); err != nil {
		return err
	}
	// Wait for cert-manager-webhook to be ready, which can take time if cert-manager
	// was re-installed after uninstalling on a cluster.
	cmd = exec.Command("kubectl", "wait", "deployment.apps/cert-manager-webhook",
		"--for", "condition=Available",
		"--namespace", "cert-manager",
		"--timeout", "5m",
	)

	_, err := Run(cmd)
	return err
}

// LoadImageToKindCluster loads a local docker image to the kind cluster
func LoadImageToKindClusterWithName(name string) error {
	cluster := "kind"
	if v, ok := os.LookupEnv("KIND_CLUSTER"); ok {
		cluster = v
	}
	kindOptions := []string{"load", "docker-image", name, "--name", cluster}
	cmd := exec.Command("kind", kindOptions...)
	_, err := Run(cmd)
	return err
}

// GetNonEmptyLines converts given command output string into individual objects
// according to line breakers, and ignores the empty elements in it.
func GetNonEmptyLines(output string) []string {
	var res []string
	elements := strings.Split(output, "\n")
	for _, element := range elements {
		if element != "" {
			res = append(res, element)
		}
	}

	return res
}
``` 

However, see that tests for the metrics and cert-manager might fit better well as e2e tests and not under the tests done using ENV TEST for the controllers. You might want to give a look at the [sample example][sdk-e2e-sample-example] implemented into [Operator-SDK][sdk] repository to know how can you do your e2e tests to ensure the basic workflows of your project.

```

File: docs/book/src/reference/envtest.md [hunk 12]
```diff
@@ -167,5 +183,43 @@ expectedOwnerReference := v1.OwnerReference{
...
+```go
+testEnv = &envtest.Environment{
+	UseExistingCluster: true,
+}
+```
+
+[metrics]: https://book.kubebuilder.io/reference/metrics.html
+[envtest]: https://pkg.go.dev/sigs.k8s.io/controller-runtime/pkg/envtest
+[setup-envtest]: https://pkg.go.dev/sigs.k8s.io/controller-runtime/tools/setup-envtest
+[cert-manager]: https://book.kubebuilder.io/cronjob-tutorial/cert-manager.html
```
Comment (camilamacedo86) [inline docs/book/src/reference/envtest.md:225]:
```suggestion
[cert-manager]: https://book.kubebuilder.io/cronjob-tutorial/cert-manager.html
[sdk-e2e-sample-example]: https://github.com/operator-framework/operator-sdk/tree/master/testdata/go/v3/memcached-operator/test/e2e
[sdk]: https://github.com/operator-framework/operator-sdk
```

**Review by camilamacedo86 — APPROVED** [review]
/approved

/ok-to-test

It shows great to me. Would be nice if we get a second reviewer as well. 

c/c @varshaprasad96 @everettraven

File: docs/book/src/reference/envtest.md [hunk 13]
```diff
@@ -76,6 +76,15 @@ err = testEnv.Stop()
 
 Logs from the test runs are prefixed with `test-env`.
 
+<aside class="note">
+<h1>Examples</h1>
+
+You can use the plugin [DeployImage](https://book.kubebuilder.io/plugins/deploy-image-plugin-v1-alpha.html) to check examples. This plugin allows users to scaffold API/Controllers to deploy and manage an Operand (image) on the cluster following the guidelines and best practices. It abstracts the complexities of achieving this goal while allowing users to customize the generated code.
+
+Therefore, you can check that a test using ENV TEST will be generated for the controller which has the purpose to ensure that the Deployment is created succeffully. You can see an example of its code implementation under the `testdata` directory with the [DeployImage](https://book.kubebuilder.io/plugins/deploy-image-plugin-v1-alpha.html) samples [here](https://github.com/kubernetes-sigs/kubebuilder/blob/v3.7.0/testdata/project-v3-with-deploy-image/controllers/busybox_controller_test.go).
```
Comment (everettraven) [inline docs/book/src/reference/envtest.md:84]: Typo:
```suggestion
Therefore, you can check that a test using ENV TEST will be generated for the controller which has the purpose to ensure that the Deployment is created successfully. You can see an example of its code implementation under the `testdata` directory with the [DeployImage](https://book.kubebuilder.io/plugins/deploy-image-plugin-v1-alpha.html) samples [here](https://github.com/kubernetes-sigs/kubebuilder/blob/v3.7.0/testdata/project-v3-with-deploy-image/controllers/busybox_controller_test.go).
```

File: docs/book/src/reference/envtest.md [hunk 14]
```diff
@@ -97,12 +106,11 @@ You can use environment variables and/or flags to specify the `kubectl`,`api-ser
 ### Environment Variables
 
 | Variable name | Type | When to use |
-| --- | :--- | :--- |
-| `USE_EXISTING_CLUSTER` | boolean | Instead of setting up a local control plane, point to the control plane of an existing cluster. |
```
Comment (everettraven) [inline docs/book/src/reference/envtest.md:101]: Does this environment variable no longer get used when using envtest?

File: docs/book/src/reference/envtest.md [hunk 15]
```diff
@@ -167,5 +183,132 @@ expectedOwnerReference := v1.OwnerReference{
...
+	for _, element := range elements {
+		if element != "" {
+			res = append(res, element)
+		}
+	}
+	
+	return res
+}
+```
+However, see that tests for the metrics and cert-manager might fit better well as e2e tests and not under the tests done using ENV TEST for the controllers. You might want to give a look at the [sample example][sdk-e2e-sample-example] implemented into [Operator-SDK][sdk] repository to know how can you do your e2e tests to ensure the basic workflows of your project.
```
Comment (everettraven) [inline docs/book/src/reference/envtest.md:300]: Grammar nit:
```suggestion
However, see that tests for the metrics and cert-manager might fit better well as e2e tests and not under the tests done using ENV TEST for the controllers. You might want to give a look at the [sample example][sdk-e2e-sample-example] implemented into [Operator-SDK][sdk] repository to know how you can write your e2e tests to ensure the basic workflows of your project.
```

**Review by everettraven — CHANGES_REQUESTED** [review]
Just a couple minor nits and a question

File: docs/book/src/reference/envtest.md (hunk 14 above)
Comment (camilamacedo86) [reply docs/book/src/reference/envtest.md:101]: HUn if it was before is good we leave that, I thought that it was added. @yashsingh74 could you please just revert the changes on the table and leave it as it was?

Comment (yashsingh74) [reply docs/book/src/reference/envtest.md:101]: No, it is no longer required, as we mentioned below that how we can use the existing cluster.

File: docs/book/src/reference/envtest.md (hunk 15 above)
Comment (yashsingh74) [reply docs/book/src/reference/envtest.md]: Done

File: docs/book/src/reference/envtest.md (hunk 14 above)
Comment (yashsingh74) [reply docs/book/src/reference/envtest.md:101]: Done

File: docs/book/src/reference/envtest.md [hunk 16]
```diff
@@ -167,5 +184,132 @@ expectedOwnerReference := v1.OwnerReference{
...
+	var res []string
+	elements := strings.Split(output, "\n")
+	for _, element := range elements {
+		if element != "" {
+			res = append(res, element)
+		}
+	}
+	
+	return res
+}
```
Comment (varshaprasad96) [inline docs/book/src/reference/envtest.md:299]: We can remove this method. It doesn't seem relevant to our explanation on how to use cert-manager and prometheus. Seems out of the way, user's can directly refer to the tests and check what this method does since it seems to be an helper.
cc: @camilamacedo86 

```suggestion
```

**Review by varshaprasad96 — COMMENTED** [review]
Thanks for the contribution @yashsingh74! Just one nit. Rest looks good!

File: docs/book/src/reference/envtest.md (hunk 16 above)
Comment (yashsingh74) [reply docs/book/src/reference/envtest.md]: Done
